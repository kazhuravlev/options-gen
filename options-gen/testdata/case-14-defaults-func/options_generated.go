// Code generated by options-gen qa-version. DO NOT EDIT.

package testcase

import (
	fmt461e464ebed9 "fmt"
	"net/http"
	"time"

	errors461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/errors"
	validator461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/validator"
)

type OptOptionsSetter func(o *Options)

func NewOptions(
	options ...OptOptionsSetter,
) Options {
	var o Options

	// Setting defaults from func
	defaultOpts := getDefaults()
	o.name = defaultOpts.name

	o.timeout = defaultOpts.timeout

	o.maxAttempts = defaultOpts.maxAttempts

	o.httpClient = defaultOpts.httpClient

	for _, opt := range options {
		opt(&o)
	}
	return o
}

func WithSomeName(opt string) OptOptionsSetter {
	return func(o *Options) {
		o.name = opt
	}
}

func WithSomeTimeout(opt time.Duration) OptOptionsSetter {
	return func(o *Options) {
		o.timeout = opt
	}
}

func WithSomeMaxAttempts(opt int) OptOptionsSetter {
	return func(o *Options) {
		o.maxAttempts = opt
	}
}

func WithSomeHttpClient(opt *http.Client) OptOptionsSetter {
	return func(o *Options) {
		o.httpClient = opt
	}
}

func (o *Options) Validate() error {
	errs := new(errors461e464ebed9.ValidationErrors)
	errs.Add(errors461e464ebed9.NewValidationError("name", _validate_Options_name(o)))
	errs.Add(errors461e464ebed9.NewValidationError("timeout", _validate_Options_timeout(o)))
	errs.Add(errors461e464ebed9.NewValidationError("maxAttempts", _validate_Options_maxAttempts(o)))
	errs.Add(errors461e464ebed9.NewValidationError("httpClient", _validate_Options_httpClient(o)))
	return errs.AsError()
}

func _validate_Options_name(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.name, "required"); err != nil {
		return fmt461e464ebed9.Errorf("field `name` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_timeout(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.timeout, "min=100ms,max=30s"); err != nil {
		return fmt461e464ebed9.Errorf("field `timeout` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_maxAttempts(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.maxAttempts, "min=1,max=10"); err != nil {
		return fmt461e464ebed9.Errorf("field `maxAttempts` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_httpClient(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.httpClient, "gt=0"); err != nil {
		return fmt461e464ebed9.Errorf("field `httpClient` did not pass the test: %w", err)
	}
	return nil
}
