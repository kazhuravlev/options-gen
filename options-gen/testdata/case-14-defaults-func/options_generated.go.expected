// Code generated by options-gen. DO NOT EDIT.
package testcase

import (
	fmt461e464ebed9 "fmt"
	"net/http"
	"time"

	errors461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/errors"
	validator461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/validator"
)

type optSomeField int8

const (
	FieldSomename        optSomeField = 0
	FieldSometimeout     optSomeField = 1
	FieldSomemaxAttempts optSomeField = 2
	FieldSomehttpClient  optSomeField = 3
)

var optSomeIsSet = [4]bool{}

type OptOptionsSetter func(o *Options)

func NewOptions(
	options ...OptOptionsSetter,
) Options {
	o := Options{}

	var empty [4]bool
	optSomeIsSet = empty

	// Setting defaults from func
	defaultOpts := getDefaults()
	o.name = defaultOpts.name
	optSomeIsSet[FieldSomename] = true
	o.timeout = defaultOpts.timeout
	optSomeIsSet[FieldSometimeout] = true
	o.maxAttempts = defaultOpts.maxAttempts
	optSomeIsSet[FieldSomemaxAttempts] = true
	o.httpClient = defaultOpts.httpClient
	optSomeIsSet[FieldSomehttpClient] = true

	for _, opt := range options {
		opt(&o)
	}
	return o
}

func WithSomeName(opt string) OptOptionsSetter {
	return func(o *Options) {
		o.name = opt
		optSomeIsSet[FieldSomename] = true
	}
}

func WithSomeTimeout(opt time.Duration) OptOptionsSetter {
	return func(o *Options) {
		o.timeout = opt
		optSomeIsSet[FieldSometimeout] = true
	}
}

func WithSomeMaxAttempts(opt int) OptOptionsSetter {
	return func(o *Options) {
		o.maxAttempts = opt
		optSomeIsSet[FieldSomemaxAttempts] = true
	}
}

func WithSomeHttpClient(opt *http.Client) OptOptionsSetter {
	return func(o *Options) {
		o.httpClient = opt
		optSomeIsSet[FieldSomehttpClient] = true
	}
}

func (o *Options) Validate() error {
	errs := new(errors461e464ebed9.ValidationErrors)
	errs.Add(errors461e464ebed9.NewValidationError("name", _validate_Options_name(o)))
	errs.Add(errors461e464ebed9.NewValidationError("timeout", _validate_Options_timeout(o)))
	errs.Add(errors461e464ebed9.NewValidationError("maxAttempts", _validate_Options_maxAttempts(o)))
	errs.Add(errors461e464ebed9.NewValidationError("httpClient", _validate_Options_httpClient(o)))
	return errs.AsError()
}

func (o *Options) IsSet(field optSomeField) bool {
	return optSomeIsSet[field]
}

func _validate_Options_name(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.name, "required"); err != nil {
		return fmt461e464ebed9.Errorf("field `name` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_timeout(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.timeout, "min=100ms,max=30s"); err != nil {
		return fmt461e464ebed9.Errorf("field `timeout` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_maxAttempts(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.maxAttempts, "min=1,max=10"); err != nil {
		return fmt461e464ebed9.Errorf("field `maxAttempts` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_httpClient(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.httpClient, "gt=0"); err != nil {
		return fmt461e464ebed9.Errorf("field `httpClient` did not pass the test: %w", err)
	}
	return nil
}
